pkg/idtools/idtools.go

```diff
@@ -0,0 +1,145 @@
+package idtools
+
+import (
+	"fmt"
+	"math"
+	"os"
+
+	"github.com/docker/docker/pkg/system"
+)
+
+// IDMap contains a single entry for user namespace range remapping. An array
+// of IDMap entries represents the structure that will be provided to the Linux
+// kernel for creating a user namespace.
+type IDMap struct {
+	ContainerID int `json:"container_id"`
+	HostID      int `json:"host_id"`
+	Size        int `json:"size"`
+}
+
+// MkdirAllAs creates a directory (include any along the path) and then modifies
+// ownership to the requested uid/gid.  If the directory already exists, this
+// function will still change ownership to the requested uid/gid pair.
+func MkdirAllAs(path string, mode os.FileMode, ownerUID, ownerGID int) error {
+	return mkdirAs(path, mode, ownerUID, ownerGID, true)
+}
+
+// MkdirAs creates a directory and then modifies ownership to the requested uid/gid.
+// If the directory already exists, this function still changes ownership
+func MkdirAs(path string, mode os.FileMode, ownerUID, ownerGID int) error {
+	return mkdirAs(path, mode, ownerUID, ownerGID, false)
+}
+
+func mkdirAs(path string, mode os.FileMode, ownerUID, ownerGID int, mkAll bool) error {
+
+	if mkAll {
+		if err := system.MkdirAll(path, mode); err != nil && !os.IsExist(err) {
+			return err
+		}
+	} else {
+		if err := os.Mkdir(path, mode); err != nil && !os.IsExist(err) {
+			return err
+		}
+	}
+	// even if it existed, we will chown to change ownership as requested
+	if err := os.Chown(path, ownerUID, ownerGID); err != nil {
```

Comment:
hmm, good question.  The semantics are normally only to affect the dir(s) being created, but if there was a case where it already existed *with content* that is improperly owned, then our particular use may break.

However, I'm thinking switching to recursive has two problems:
 - performance: when the daemon starts, the root graph directory (and then graph directories for the storage driver) all come through this code path, even though after the first time these all exist.  They may have significant amounts of content over time, so recursive would be painful, and..
 - functionally wrong for the brute force case: if **/var/lib/docker/2000.2000/aufs** was recursively `chown`'ed the content (e.g. image layers) that is *correctly* owned by other uid/gid pairs than root would all change to remapped root.  Instead the tree would have to be filtered by root owned content only, if recursive is the right answer at all.